server/src/main/java/com/cloud/api/query/QueryManagerImpl.java

```diff
@@ -1447,13 +1447,18 @@ private Pair<List<Long>, Integer> searchForUserVMIdsAndCount(ListVMsCmd cmd) {
         }
 
         if (keyword != null) {
-            userVmSearchBuilder.and().op("keywordDisplayName", userVmSearchBuilder.entity().getDisplayName(), Op.LIKE);
-            userVmSearchBuilder.or("keywordName", userVmSearchBuilder.entity().getHostName(), Op.LIKE);
-            userVmSearchBuilder.or("keywordState", userVmSearchBuilder.entity().getState(), Op.EQ);
+            SearchBuilder<UserVmJoinVO> userVmJoinSearchBuilder = _userVmJoinDao.createSearchBuilder();
```

Comment:
It would be better to make a join with NIC table instead of UserVMJoin.

Reply:
Also, `ip4_address` column in `nics` table might need an index as well.

Reply:
this would involve 3 different tables. Do you think it would still be beneficial? I tried something along these lines:
```
SearchBuilder<NicVO> nicsSearch = nicDao.createSearchBuilder();
nicsSearch.or("keywordIpAddress", nicsSearch.entity().getIPv4Address(), Op.LIKE);
nicsSearch.or("keywordIp6Address", nicsSearch.entity().getIPv6Address(), Op.LIKE);
userVmSearchBuilder.join("nicIpKeyword", nicsSearch, nicsSearch.entity().getInstanceId(), userVmSearchBuilder.entity().getId(), JoinBuilder.JoinType.INNER);

SearchBuilder<IPAddressVO> publicIPAddressSearch = ipAddressDao.createSearchBuilder();
publicIPAddressSearch.or("keywordPublicIpAddress", publicIPAddressSearch.entity().getAddress(), Op.LIKE);
userVmSearchBuilder.join("userIpKeyword", publicIPAddressSearch, publicIPAddressSearch.entity().getAssociatedWithVmId(), userVmSearchBuilder.entity().getId(), JoinBuilder.JoinType.INNER);
```

But this doesn't seem to work as expected, as each  inner join does an AND op with the keyword. Any pointers on what could be missing?

Reply:
I haven't tested this. But the code will look something like this:

```java
  if (keyword != null) {
            userVmSearchBuilder.and().op("keywordDisplayName", userVmSearchBuilder.entity().getDisplayName(), Op.LIKE);
            userVmSearchBuilder.or("keywordName", userVmSearchBuilder.entity().getHostName(), Op.LIKE);
            userVmSearchBuilder.or("keywordState", userVmSearchBuilder.entity().getState(), Op.EQ);
            if (isRootAdmin) {
                userVmSearchBuilder.or("keywordInstanceName", userVmSearchBuilder.entity().getInstanceName(), Op.LIKE );
            }

            SearchBuilder<IPAddressVO> ipAddressSearch = ipAddressDao.createSearchBuilder();
            userVmSearchBuilder.join("ipAddressSearch", ipAddressSearch, ipAddressSearch.entity().getAssociatedWithVmId(), userVmSearchBuilder.entity().getId(), JoinBuilder.JoinType.LEFT);

            SearchBuilder<NicVO> nicSearch = nicDao.createSearchBuilder();
            userVmSearchBuilder.join("nicSearch", nicSearch, JoinBuilder.JoinType.LEFT,
                    JoinBuilder.JoinCondition.AND,
                    nicSearch.entity().getInstanceId(), userVmSearchBuilder.entity().getId(),
                    nicSearch.entity().getRemoved(), userVmSearchBuilder.entity().setLong(null));

            userVmSearchBuilder.or("ipAddressSearch", "publicIpSearch", ipAddressSearch.entity().getAddress(), Op.LIKE);

            userVmSearchBuilder.or("nicSearch", "ip4Address", nicSearch.entity().getIPv4Address(), Op.LIKE);
            userVmSearchBuilder.or("nicSearch", "ip6Address", nicSearch.entity().getIPv6Address(), Op.LIKE);

            userVmSearchBuilder.cp();
        }

```

and for setting the paramerts
```java
if (keyword != null) {
            String keywordMatch = "%" + keyword + "%";
            userVmSearchCriteria.setParameters("keywordDisplayName", keywordMatch);
            userVmSearchCriteria.setParameters("keywordName", keywordMatch);
            userVmSearchCriteria.setParameters("keywordState", keyword);
            if (isRootAdmin) {
                userVmSearchCriteria.setParameters("keywordInstanceName", keywordMatch);
            }

            userVmSearchCriteria.setParameters("publicIpSearch", keywordMatch);
            userVmSearchCriteria.setParameters("ip4Address", keywordMatch);
            userVmSearchCriteria.setParameters("ip6Address", keywordMatch);
        }
```

Reply:
Thanks vishesh, wasn't aware of this new addition. I've updated the code to incorporate joins and not use the view